### PR TITLE
(Feature) G$ Decimals should be dynamic based on token 

### PR DESCRIPTION
**Pull request by johnsmith-gooddollar**
# Description

- [x] updated G$ tokens/currencies definition depending on chain id
- [x] added helper hooks to format G$/GOOD/GDX values
- [ ] TODO (@L03TJ3 , most probably at https://github.com/GoodDollar/GoodProtocolUI) Make sure all inputs/display components are using `packages/good-design/src/core/inputs/TokenOutput.tsx` and `packages/good-design/src/core/inputs/TokenInput.tsx` from `gooddesign`

About [#284](https://github.com/GoodDollar/GoodProtocolUI/issues/284)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

**Comment by johnsmith-gooddollar** [comment]
@L03TJ3 It seems that's all we had to change.
I've found `TokenInput`/`TokenOutput` components formatting numbers according to the decimals they're receiving as prop. So there's no need to create basic components for display and input - they're already exists.

We only need to make sure any token-related field (or textbox) at GoodSwap uses such components and passes the correct token/decimals

**Comment by johnsmith-gooddollar** [comment]
**UPD**

1. adde some helpers to format G$/GOOD/GDX values
2. I've checked how `TokenInput`/`TokenOutput` are used now, have some questions to discuss, let's do it next Mon

**Comment by sirpy** [comment]
@johnsmith-gooddollar @L03TJ3 
token decimals should be read from the contract, not hard coded

**Comment by johnsmith-gooddollar** [comment]
> @johnsmith-gooddollar @L03TJ3
> token decimals should be read from the contract, not hard coded

@L03TJ3 please assist here. I'll prepare React-side part, you'll need to implement contract call

**Comment by L03TJ3** [comment]
Basic implementation (without using Token Input / Output from gooddesign)
can be found in PR for savings UI:  [282](https://github.com/GoodDollar/GoodProtocolUI/pull/282)

**Review by johnsmith-gooddollar — COMMENTED** [review]
not as I wanted, need to revert some changes

File: packages/good-design/src/apps/bridge/MicroBridge.tsx [hunk 1]
```diff
@@ -187,7 +187,7 @@ export const MicroBridge = ({
       </Flex>
       <Flex direction="column" alignItems="flex-start" justifyContent="flex-start" width="100%">
         <Text fontFamily="subheading" bold color="lightGrey:alpha.80">ENTER AMOUNT</Text>
-        <TokenInput balanceWei={balanceWei} decimals={2} onChange={setInput} minAmountWei={minAmountWei} />
+        <TokenInput token="G$" balanceWei={balanceWei} onChange={setInput} minAmountWei={minAmountWei} />
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/apps/bridge/MicroBridge.tsx:190]:
```suggestion
        <TokenInput balanceWei={balanceWei} onChange={setInput} minAmountWei={minAmountWei} />
```

File: packages/good-design/src/apps/bridge/MicroBridge.tsx [hunk 2]
```diff
@@ -203,7 +203,7 @@ export const MicroBridge = ({
           >
             You will receive on {targetChain} 
           </Text>
-        <TokenOutput outputValue={expectedToReceive ?? '0'} />
+        <TokenOutput token="G$" outputValue={expectedToReceive ?? '0'} />
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/apps/bridge/MicroBridge.tsx:206]:
```suggestion
        <TokenOutput outputValue={expectedToReceive ?? '0'} />
```

@L03TJ3 you have token = G$ as default prop

**Comment by johnsmith-gooddollar** [comment]
@L03TJ3 updated PR please check

File: packages/good-design/src/core/layout/BalanceGD.tsx [hunk 3]
```diff
@@ -26,9 +26,8 @@ const BalanceView: FC<Required<BalanceGDProps> & { amount: CurrencyValue }> = me
 
 const BalanceGD: FC<BalanceGDProps> = ({ gdPrice }) => {
   const { G$ } = useG$Balance("everyBlock");
```
Comment (sirpy) [inline packages/good-design/src/core/layout/BalanceGD.tsx:28]: that's very expensive, dont do every block ever

File: packages/sdk-v2/src/sdk/base/react.ts [hunk 4]
```diff
@@ -191,39 +207,24 @@ export function useG$Balance(refresh: QueryParams["refresh"] = "never") {
...
+    { refresh: refreshOrNever, chainId: MAINNET as unknown as ChainId }
+  );
+
+  const [g$Value, goodValue, gdxValue] = [...results, mainnetGdx].map(
+    result => result && !result.error ? result.value[0] : undefined
   );
+  
+  const g$Balance = <CurrencyValue>useG$Amount(g$Value);
+  const goodBalance = <CurrencyValue>useG$Amount(goodValue, "GOOD");
+  const gdxBalance = <CurrencyValue>useG$Amount(gdxValue, "GDX");
```
Comment (sirpy) [inline packages/sdk-v2/src/sdk/base/react.ts:221]: what is this syntax?
usually if you want to cast you use "as CurrencyValue"

File: packages/sdk-v2/src/sdk/savings/react.ts [hunk 5]
```diff
@@ -131,6 +131,10 @@ export const useSavingsStats = (requiredChainId: number, refresh: QueryParams["r
     apy: undefined
   };
 
+  const [, totalStaked, totalRewardsPaid] = results[0]?.value ?? []; // eslint-disable-line no-unsafe-optional-chaining
+  const staked = useG$Amount(totalStaked);
```
Comment (sirpy) [inline packages/sdk-v2/src/sdk/savings/react.ts:135]: you need to verify that EVERYWHERE you use useG$Amount you also pass the requiredChainId from the surrounding hook if relevant

**Review by sirpy — CHANGES_REQUESTED** [review]

File: packages/good-design/src/core/inputs/TokenOutput.tsx [hunk 6]
```diff
@@ -1,26 +1,33 @@
+import { G$Token, useG$Decimals } from "@gooddollar/web3sdk-v2";
 import { Box, Input } from "native-base";
 import React from "react";
 import { NumericFormat } from "react-number-format";
 
 export const TokenOutput = ({
   outputValue,
+  token = 'G$',
```
[the remark this answers is not in the document]
Comment (johnsmith-gooddollar) [reply packages/good-design/src/core/inputs/TokenOutput.tsx:8]: We had two options before (token +decimals), at least one of them had to be provided

will revert to this

File: packages/sdk-v2/src/sdk/base/react.ts (hunk 4 above)
Comment (johnsmith-gooddollar) [reply packages/sdk-v2/src/sdk/base/react.ts:221]: Just alternative cast syntax

File: packages/good-design/src/core/layout/BalanceGD.tsx [hunk 7]
```diff
@@ -2,18 +2,25 @@ import React, { FC, memo } from "react";
...
 import { Fraction } from "@uniswap/sdk-core";
-import { CurrencyValue } from "@usedapp/core";
+import { CurrencyValue, QueryParams } from "@usedapp/core";
 
 interface BalanceGDProps {
   gdPrice?: Fraction;
+  refresh?: QueryParams["refresh"];
 }
 
+const BalanceGD: FC<BalanceGDProps> = ({ gdPrice, refresh = "never" }) => {
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/core/layout/BalanceGD.tsx:12]: let's add **optional** `requiredChainId?: number` here too

```suggestion
const BalanceGD: FC<BalanceGDProps> = ({ gdPrice, requiredChainId, refresh = "never" }) => {
```

File: packages/good-design/src/core/layout/BalanceGD.tsx [hunk 8]
```diff
@@ -2,18 +2,25 @@ import React, { FC, memo } from "react";
 import { Text, View } from "native-base";
 import { useG$Balance } from "@gooddollar/web3sdk-v2";
 import { Fraction } from "@uniswap/sdk-core";
-import { CurrencyValue } from "@usedapp/core";
+import { CurrencyValue, QueryParams } from "@usedapp/core";
 
 interface BalanceGDProps {
   gdPrice?: Fraction;
+  refresh?: QueryParams["refresh"];
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/core/layout/BalanceGD.tsx:9]:
```suggestion
  refresh?: QueryParams["refresh"];
  requiredChainId?: number;
```

File: packages/good-design/src/core/layout/BalanceGD.tsx [hunk 9]
```diff
@@ -2,18 +2,25 @@ import React, { FC, memo } from "react";
...
 }
 
+const BalanceGD: FC<BalanceGDProps> = ({ gdPrice, refresh = "never" }) => {
+  const { G$ } = useG$Balance(refresh);
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/core/layout/BalanceGD.tsx:13]:
```suggestion
  const { G$ } = useG$Balance(refresh, requiredChainId);
```

**Pull request merged**